File: 0049-group-anagrams/0049-group-anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        int m = strs.size();
        int n = strs[0].size();
        unordered_map<string, vector<int>> mp;
        
        for (int i = 0; i < m; ++i) {
            string temp = strs[i];
            sort(temp.begin(), temp.end());
            mp[temp].push_back(i);
        }
        vector<vector<string>> res;
        for (auto a : mp) {
            vector<string> temp;
            for (int i = 0; i < a.second.size(); ++i) {
                temp.push_back(strs[a.second[i]]);
            }
            res.push_back(temp);
        }
        return res;
        
    }
};

File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        int n = strs.length;
        HashMap<String, ArrayList<String>> mp = new HashMap<>();
        for (String str: strs) {
            char[] arr = str.toCharArray();
            Arrays.sort(arr);
            String sortedStr = String.valueOf(arr);
            if (!mp.containsKey(sortedStr)) {
                mp.put(sortedStr, new ArrayList<>());
            }
            mp.get(sortedStr).add(str);
        }
        List<List<String>> res = new ArrayList<>();
        for (Map.Entry<String, ArrayList<String>> entry: mp.entrySet()) {
            res.add(entry.getValue());
        }
        return res;
    }
}